### PR TITLE
feat: add support for TurboModule

### LINKED ISSUE #836: Add support for TurboModule
Following the flag to enable Fabric (#785), we should also add a flag to enable TurboModules:

- [x] [Enable TurboModule on Android](https://reactnative.dev/docs/next/new-architecture-app-modules-android)
- [x] [Enable TurboModule on iOS](https://reactnative.dev/docs/next/new-architecture-app-modules-ios)
- [x] Hook up autolinking

### Blocked by

- ~~Autolinking support on Android: https://github.com/reactwg/react-native-new-architecture/discussions/2#discussioncomment-2437670~~ Implemented in https://github.com/react-native-community/cli/pull/1603

### PR DESCRIPTION
**Pull request by tido64**
### Description

Adds support for TurboModule. I'm still seeing issues with Gradle not building things in the correct order. Until a new version of Gradle is released, we will have to point people at [Android NDK: ERROR:/~/android/src/main/jni/Android.mk:fb: LOCAL_SRC_FILES points to a missing file](https://github.com/microsoft/react-native-test-app/wiki/Troubleshooting#android-ndk-errorandroidsrcmainjniandroidmkfb-local_src_files-points-to-a-missing-file).

Resolves #836.

### Blocked by

- ~~Autolinking support on Android: https://github.com/reactwg/react-native-new-architecture/discussions/2#discussioncomment-2437670~~ Implemented in https://github.com/react-native-community/cli/pull/1603

### Platforms affected

- [x] Android
- [x] iOS
- [ ] macOS
- [ ] Windows

### Test plan

- Use `react-native` 0.69: `yarn set-react-version 0.69`
- Use `@react-native-community/cli` 9.0:
   ```diff
   diff --git a/package.json b/package.json
   index c30ed18..792e470 100644
   --- a/package.json
   +++ b/package.json
   @@ -138,6 +138,9 @@
      "resolutions": {
        "@commitlint/is-ignored/semver": "^7.3.5",
        "@microsoft/eslint-plugin-sdl/eslint-plugin-react": "^7.26.0",
   +    "@react-native-community/cli": "^9.0.0-0",
   +    "@react-native-community/cli-platform-android": "^9.0.0-0",
   +    "@react-native-community/cli-platform-ios": "^9.0.0-0",
        "@react-native-community/cli-tools/shell-quote": "^1.7.3",
        "core-js-compat/semver": "^7.3.5",
        "npm/chalk": "^4.1.0"
   ```
- Re-run `yarn` after making the changes above
- Android:
  - Enable TurboModule:
    ```diff
    diff --git a/example/android/gradle.properties b/example/android/gradle.properties
    index 3cd810b..b19e049 100644
    --- a/example/android/gradle.properties
    +++ b/example/android/gradle.properties
    @@ -36,7 +36,7 @@ android.enableJetifier=true

     # Enable new architecture, i.e. Fabric + TurboModule - implies USE_FABRIC=1.
     # Note that this is incompatible with web debugging.
    -#newArchEnabled=true
    +newArchEnabled=true

     # Uncomment the line below if building react-native from source
     #ANDROID_NDK_VERSION=21.4.7075529
    ```
  - Build: `yarn android`
- iOS:
  - Install Pods: `RCT_NEW_ARCH_ENABLED=1 pod install`
  - Build: `yarn ios`

### CODE REVIEW

**Comment by kelset** [comment]
couple of quick things about the test steps:
* you forgot to mention that after the changes in the package.json in the example folder, one has to re-run yarn ;)
* also, technically, you can also turn on the modules in the Podfile via the flag `turbomodule_enabled` 🤓

Now, talking about testing the example app:
* on Android, for some reason the App(Modal) doesn't work - I don't know if that's expected
* also, despite me only turning on the TM flag, the screen says I've also enabled Fabric, see screenshot:
Screenshot 2022-07-05 at 12 23 56

* it took 28 mins to build on my maxed intel Macbook pro 16', but I assume that's expected because Hermes needs to be built from source
* on iOS, everything seems to work fine (both normal way and modal way) but again it tells me I've enabled Fabric:
Screenshot 2022-07-05 at 12 37 51

**Comment by tido64** [comment]
> * you forgot to mention that after the changes in the package.json in the example folder, one has to re-run yarn ;)

Oops… description should be updated now.

> * also, technically, you can also turn on the modules in the Podfile via the flag `turbomodule_enabled` 🤓

Correct. I wanted a focus on the flags that are the same in core. Less to teach.

> * on Android, for some reason the App(Modal) doesn't work - I don't know if that's expected

This shouldn't be broken. I'll have a look.

> * also, despite me only turning on the TM flag, the screen says I've also enabled Fabric, see screenshot:

This is expected. TM depends on code that is under the Fabric flag.

> * it took 28 mins to build on my maxed intel Macbook pro 16', but I assume that's expected because Hermes needs to be built from source

Unfortunately, yes.

> * on iOS, everything seems to work fine (both normal way and modal way) but again it tells me I've enabled Fabric:

This is expected. TM depends on code that is under the Fabric flag.

**Comment by tido64** [comment]
> > * on Android, for some reason the App(Modal) doesn't work - I don't know if that's expected
> 
> This shouldn't be broken. I'll have a look.
> 

Did you test on a device or emulator? Which version of Android are you running? It works for me here:

Screenshot_1657031749

This is what I'm running:

image

**Comment by kelset** [comment]
Tested on emulator, this device:
Screenshot 2022-07-05 at 16 20 22

I can try to update Android Studio and clean it all up, and do from scratch then report back

**Comment by kelset** [comment]
Just tested the branch once again - everything is working as intended:
Screenshot 2022-07-22 at 15 49 42

File: example/android/gradle.properties [hunk 1]
```diff
@@ -34,5 +34,9 @@ android.enableJetifier=true
 # Enable Fabric at runtime.
 #USE_FABRIC=1
 
+# Enable new architecture, i.e. Fabric + TurboModule - implies USE_FABRIC=1.
+# Note that this is incompatible with web debugging.
+#newArchEnabled=true
```
Comment (kelset) [inline example/android/gradle.properties:39]:
```suggestion
# newArchEnabled=true
```
nit - sorry it was bothering me a lot 🤣

File: ios/use_react_native-0.68.rb [hunk 2]
```diff
@@ -20,6 +14,29 @@ def include_react_native!(options)
...
 
+  return unless new_arch_enabled
+
+  Pod::UI.warn(
+    'As of writing, TurboModule is still experimental and subject to change. ' \
+    'For more information, please see ' \
+    'https://reactnative.dev/docs/next/new-architecture-app-modules-ios.'
+  )
+  # At the moment, Fabric and TurboModule code are intertwined. We need to
+  # enable Fabric for some code that TurboModule relies on.
```
Comment (kelset) [inline ios/use_react_native-0.68.rb:25]: really? I thought they were independent - is this something that we bubbled to Meta?

**Review by kelset — APPROVED** [review]
:shipit:

File: example/android/gradle.properties (hunk 1 above)
Comment (tido64) [reply example/android/gradle.properties:39]: This line isn't a comment though. It's a setting that's commented out 😄

File: ios/use_react_native-0.68.rb (hunk 2 above)
Comment (tido64) [reply ios/use_react_native-0.68.rb:25]: I think they know, and probably the reason why they only have the one flag instead of two.

**Pull request merged**